sound/soc/sof/pm.c

```diff
@@ -429,12 +429,61 @@ EXPORT_SYMBOL(snd_sof_runtime_resume);
 
 int snd_sof_resume(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	int ret;
+
+	if (acpi_target_system_state() == ACPI_STATE_S0) {
+		dev_dbg(sdev->dev, "ADSP will exit from D0i3...\n");
+		ret = snd_sof_set_d0_state(sdev, SOF_DSP_D0I0);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed to exit from D0I3 %d\n",
+				ret);
+			return ret;
+		}
+
+		pci_restore_state(pci);
+		disable_irq_wake(pci->irq);
+
+		/* resume DMA trace, only need send ipc */
+		ret = snd_sof_init_trace_ipc(sdev);
+		if (ret < 0) {
+			/* non fatal */
+			dev_warn(sdev->dev,
+				 "warning: failed to init trace after resume %d\n",
+				 ret);
+		}
+
+		return 0;
+	}
+
 	return sof_resume(dev, false);
 }
 EXPORT_SYMBOL(snd_sof_resume);
 
 int snd_sof_suspend(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	int ret;
+
+	if (acpi_target_system_state() == ACPI_STATE_S0) {
```

Comment:
this is confusing. You check for the same state in suspend and resume?

Reply:
correct, for resuming, I should not use this, let me change.